### ISSUE TITLE
import breaks on a IFC which has missing entrys

### ISSUE THREAD

**Issue by berndhahnebach** [yorikvanhavre/FreeCAD-NativeIFC#23]
On a huge ifc from a architect I got problems. I have been used this ifc on various viewers including Solibrie, BIMCollabZoom, BIMVision and FreeCAD and BlenderBIM, as well as CAD Allplan. Noone returned big problems and all parts seams correct, BUT

dozens of objects missing the ObjectPlacement. Somehow all the programme did not have a problem with it. As an example some code attached ... 

```
Python 3.10.9 | packaged by conda-forge | (main, Feb  2 2023, 20:14:58) [MSC v.1929 64 bit (AMD64)] on win32
Type 'help', 'copyright', 'credits' or 'license' for more information.
>>> 
>>> ### Begin command Std_Open
>>> import importIFCparse
>>> importIFCparse.open(u"C:/Users/BHA/Desktop/fehlerhaftes_ifc_laed_nicht_nativ_ifc/4130_ARC_ROHM01_20230203.ifc")
>>> import ifcopenshell
>>> ifcfile = ifc = ifcopenshell.open("C:/Users/BHA/Desktop/fehlerhaftes_ifc_laed_nicht_nativ_ifc/4130_ARC_ROHM01_20230203.ifc")
>>> print(ifc.by_type("IfcProject")[0])
>>> ### End command Std_Open
>>> 
>>> ifcfile[409407]
#409407=IfcSlab('1GQsvNK6n9e9nfjgTH$n8Z',#11,'Beton','Dach, 22 cm Stärke
22 Beton','Dach 22cm',#395870,#409405,$,.ROOF.)
>>> 
>>> ifcfile[409407].ObjectPlacement
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\entity_instance.py", line 136, in __getattr__
    self.wrapped_data.get_argument(idx), self.wrapped_data.file
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\ifcopenshell_wrapper.py", line 4833, in get_argument
    return _ifcopenshell_wrapper.entity_instance_get_argument(self, *args)
RuntimeError: Instance #395870 not found
>>> 
>>> ifcfile[395870]
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\file.py", line 313, in __getitem__
    return entity_instance(self.wrapped_data.by_id(key), self)
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\ifcopenshell_wrapper.py", line 4515, in by_id
    return _ifcopenshell_wrapper.file_by_id(self, id)
RuntimeError: Instance #395870 not found
>>>
``` 

the native IFC importer has problems with this or at least he is the only one who reports back the big problems ... But the representation of thes objects seams shown in 3Dview. IMHO we should inform the user but catch such problem.

**Comment by berndhahnebach** [comment]
```
10:25:47  ERROR:root:unhandled exception occurred setting attribute name for #60386=IfcSlab('0uOv8$f1f948AnmH38W1bV',#11,'Beton','Geschossdecke mit Dach, 41 cm Stärke
Beton 35 cm
Dämmung 6 cm','Dach Überstand EG 41-34cm',#60109,#60384,$,.ROOF.)
Traceback (most recent call last):
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\entity_instance.py", line 477, in _
    attr_value = self[i]
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\entity_instance.py", line 242, in __getitem__
    self.wrapped_data.get_argument(key), self.wrapped_data.file
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\ifcopenshell_wrapper.py", line 4833, in get_argument
    return _ifcopenshell_wrapper.entity_instance_get_argument(self, *args)
RuntimeError: Instance #60109 not found
10:25:47  ERROR:root:unhandled exception occurred setting attribute name for #69084=IfcSlab('3r4oWN7Ir0DvAASUVQBnIK',#11,'Beton','EG Geschossdecke, 36 cm Stärke
Betondecke 30 cm
Dämmung 6 cm','Dach über TG unter Hof 36cm',#68975,#69082,$,.ROOF.)
Traceback (most recent call last):
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\entity_instance.py", line 477, in _
    attr_value = self[i]
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\entity_instance.py", line 242, in __getitem__
    self.wrapped_data.get_argument(key), self.wrapped_data.file
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\ifcopenshell_wrapper.py", line 4833, in get_argument
    return _ifcopenshell_wrapper.entity_instance_get_argument(self, *args)
RuntimeError: Instance #68975 not found

...
hundreds of these errors
...

#15597=IfcSlab('3K3$EnLb15KBFDjsg3NhM_',#11,'Beton','Bodenplatte, 40 cm Stärke
Betondecke 35 cm
Überzug 5 cm','Bodenplatte UG 40cm_G',#14600,#15595,$,.BASESLAB.)
Traceback (most recent call last):
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\entity_instance.py", line 477, in _
    attr_value = self[i]
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\entity_instance.py", line 242, in __getitem__
    self.wrapped_data.get_argument(key), self.wrapped_data.file
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\ifcopenshell_wrapper.py", line 4833, in get_argument
    return _ifcopenshell_wrapper.entity_instance_get_argument(self, *args)
RuntimeError: Instance #14600 not found
10:25:47  ERROR:root:unhandled exception occurred setting attribute name for #15714=IfcSlab('161ixsdHX0SuzHkCoExr8C',#11,'Beton','Bodenplatte, 40 cm Stärke
Betondecke 35 cm
Überzug 5 cm','Bodenplatte UG 40cm_G',#15629,#15712,$,.BASESLAB.)
Traceback (most recent call last):
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\entity_instance.py", line 477, in _
    attr_value = self[i]
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\entity_instance.py", line 242, in __getitem__
    self.wrapped_data.get_argument(key), self.wrapped_data.file
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\ifcopenshell_wrapper.py", line 4833, in get_argument
    return _ifcopenshell_wrapper.entity_instance_get_argument(self, *args)
RuntimeError: Instance #15629 not found
10:25:55  DEBUG: ifc_tools.get_coin: Invalid iterator
10:26:14  DEBUG: ifc_tools.get_coin: Invalid iterator
10:26:16  DEBUG: ifc_tools.get_coin: Invalid iterator
10:26:16  DEBUG: ifc_tools.get_coin: Invalid iterator

...
hundreds of these errors
...

10:28:24  DEBUG: ifc_tools.get_coin: Invalid iterator
10:28:28  DEBUG: ifc_tools.get_coin: Invalid iterator
10:28:28  DEBUG: ifc_tools.get_coin: Invalid iterator
10:28:28  DEBUG: ifc_tools.get_coin: Invalid iterator
10:28:28  Recompute failed! Please check report view.
10:28:28  Imported 4130_ARC_ROHM01_20230203.ifc ( 22.86 Mb ) in 02:51
```

**Comment by berndhahnebach** [comment]
![screen1](https://user-images.githubusercontent.com/6129869/221143901-716a5401-b5ee-4d7c-a570-fd8d0841ddbd.png)

**Comment by berndhahnebach** [comment]
Attached a IFC which I did edit to get similar behaviour. ... For some reason the tree view does not show the recompute problem mark on this object.


```
10:40:11  ERROR:root:unhandled exception occurred setting attribute name for #80=IfcSlab('2JwE5lcKXFKvWOrg3uyxKD',#5,'Structure','',$,#77,#79,$,.NOTDEFINED.)
Traceback (most recent call last):
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\entity_instance.py", line 477, in _
    attr_value = self[i]
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\entity_instance.py", line 242, in __getitem__
    self.wrapped_data.get_argument(key), self.wrapped_data.file
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\ifcopenshell_wrapper.py", line 4833, in get_argument
    return _ifcopenshell_wrapper.entity_instance_get_argument(self, *args)
RuntimeError: Instance #77 not found
10:40:11  Imported no_placement_edited.ifc ( 0.0 Mb ) in 00:00
```



[no_placement_edited.ifc.txt](https://github.com/yorikvanhavre/FreeCAD-NativeIFC/files/10822753/no_placement_edited.ifc.txt)


![grafik](https://user-images.githubusercontent.com/6129869/221145779-7e7c50d2-8dfb-48e0-ab88-bb4d39aaefdf.png)

**Comment by yorikvanhavre** [comment]
This looks like it happens fully inside ifcopenshell... I can reproduce the error with python + ifcopenshell only:
```python
>>> import ifcopenshell
>>> f=ifcopenshell.open("Downloads/no_placement_edited.ifc")
>>> f.by_type("IfcSlab")
[#80=IfcSlab('2JwE5lcKXFKvWOrg3uyxKD',#5,'Structure','',$,#77,#79,$,.NOTDEFINED.)]
>>> f[77]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/yorik/.local/lib/python3.11/site-packages/ifcopenshell/file.py", line 313, in __getitem__
    return entity_instance(self.wrapped_data.by_id(key), self)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/yorik/.local/lib/python3.11/site-packages/ifcopenshell/ifcopenshell_wrapper.py", line 3377, in by_id
    return _ifcopenshell_wrapper.file_by_id(self, id)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Instance #77 not found
```
It seems to me like wrong IFC formatting to have #77 linked to, but the linked object does not exist. However, indeed BlenderBIM opens it without errors. So I suppose BlenderBIM does some magic here... @Moult certainly knows? ;)

**Comment by berndhahnebach** [comment]
yep fully inside ifcopenshell and I am able to reproduce this in ifcopenshell only as well with the monster file.

Have you tried normal FreeCAD importer? No warning nothing from ifcopenshell ... 


```
11:01:46  Opening 'C:/Users/BHA/Desktop/fehlerhaftes_ifc_laed_nicht_nativ_ifc/no_placement_edited.ifc'... done.
11:01:46  Building types and relationships table...
11:01:46  done.
11:01:46  Parsing 4 BIM objects...
11:01:46  1 / 4 object #80 : IfcSlab no layer found IfcSlab (arch) 2k 1 solids : ID80 Structure 
11:01:46  2 / 4 object #81 : IfcSite (arch) skipped.
11:01:46  3 / 4 object #83 : IfcBuilding (arch) skipped.
11:01:46  4 / 4 object #85 : IfcBuildingStorey (arch) skipped.
11:01:46  Processing Struct relationships...done
11:01:46  Processing Arch relationships...done.
11:01:46  Finished importing 0.0 MB in 2 seconds, or 497 s/MB

```


BTW: The monster big file opens in BlenderBIM and BlenderBIM does not complain. At least not in a way I have noticed it.

**Comment by berndhahnebach** [comment]
ifcopenshell validate finds it ...
```

>>> ### Begin command Std_Open
>>> import importIFCparse
>>> importIFCparse.open(u"C:/Users/BHA/Desktop/fehlerhaftes_ifc_laed_nicht_nativ_ifc/no_placement_edited.ifc")
>>> ifcfile = ifc = ifcopenshell.open("C:/Users/BHA/Desktop/fehlerhaftes_ifc_laed_nicht_nativ_ifc/no_placement_edited.ifc")
>>> print(ifc.by_type("IfcProject")[0])
>>> Gui.SendMsgToActiveView("ViewFit")
>>> ### End command Std_Open
>>> 
>>> json_logger = ifcopenshell.validate.json_logger()
>>> ifcopenshell.validate.validate(ifcfile, json_logger)
>>> 
>>> json_logger.statements
[{'level': 'error', 'message': 'Invalid attribute value', 'type': 'schema', 'instance': #80=IfcSlab('2JwE5lcKXFKvWOrg3uyxKD',#5,'Structure','',$,#77,#79,$,.NOTDEFINED.), 'attribute': 'IfcSlab.ObjectPlacement'}]
>>> 
>>> 
```

**Comment by yorikvanhavre** [comment]
It's really strange since the standard ImportIFC also uses the same ifcopenshell iterator... Maybe because inside we are querying object transformations? I'll test that...

**Comment by berndhahnebach** [comment]
open ifc with latest native ifc module and copy code ... 

AFIAK this is what happens in importIFC ...

```
import ifcopenshell, os, Part
from ifcopenshell import geom
settings = ifcopenshell.geom.settings()
settings.set(settings.USE_BREP_DATA,True)
settings.set(settings.SEW_SHELLS,True)
settings.set(settings.USE_WORLD_COORDS,True)


p=ifcfile[80]
r = p.Representation
print(p, '\n', r)
cr = ifcopenshell.geom.create_shape(settings, p)
brep = cr.geometry.brep_data
shape = Part.Shape()
shape.importBrepFromString(brep)
Part.show(shape)
```

no error here ... Shape shows in FreeCAD ...

```
Python 3.10.9 | packaged by conda-forge | (main, Feb  2 2023, 20:14:58) [MSC v.1929 64 bit (AMD64)] on win32
Type 'help', 'copyright', 'credits' or 'license' for more information.
>>> import ifcopenshell, os, Part
>>> from ifcopenshell import geom
>>> settings = ifcopenshell.geom.settings()
>>> settings.set(settings.USE_BREP_DATA,True)
>>> settings.set(settings.SEW_SHELLS,True)
>>> settings.set(settings.USE_WORLD_COORDS,True)
>>> 
>>> 
>>> p=ifcfile[80]
>>> r = p.Representation
>>> print(p, '\n', r)
#80=IfcSlab('2JwE5lcKXFKvWOrg3uyxKD',#5,'Structure','',$,#77,#79,$,.NOTDEFINED.) 
 #79=IfcProductDefinitionShape($,$,(#78))
>>> cr = ifcopenshell.geom.create_shape(settings, p)
>>> brep = cr.geometry.brep_data
>>> shape = Part.Shape()
>>> shape.importBrepFromString(brep)
>>> Part.show(shape)
<Part::PartFeature>
>>> 
```

**Comment by yorikvanhavre** [comment]
This might be the issue: https://github.com/yorikvanhavre/FreeCAD-NativeIFC/blob/main/ifc_tools.py#L467

**Comment by berndhahnebach** [comment]
> This might be the issue: https://github.com/yorikvanhavre/FreeCAD-NativeIFC/blob/main/ifc_tools.py#L467

No it is not!

It is this line ... https://github.com/yorikvanhavre/FreeCAD-NativeIFC/blob/e7bd20e6eedf3afd703ad122cc9996a7747b6a82/ifc_tools.py#L199

open ifc and run

`ifcfile[80].get_info().items()`

will result in 

```

15:41:54  ERROR:root:unhandled exception occurred setting attribute name for #80=IfcSlab('2JwE5lcKXFKvWOrg3uyxKD',#5,'Structure','',$,#77,#79,$,.NOTDEFINED.)
Traceback (most recent call last):
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\entity_instance.py", line 477, in _
    attr_value = self[i]
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\entity_instance.py", line 242, in __getitem__
    self.wrapped_data.get_argument(key), self.wrapped_data.file
  File "C:\0_BHA_privat\progr\FreeCAD\FreeCAD_0.21.xxxxx\bin\lib\site-packages\ifcopenshell\ifcopenshell_wrapper.py", line 4833, in get_argument
    return _ifcopenshell_wrapper.entity_instance_get_argument(self, *args)
RuntimeError: Instance #77 not found
```

**Comment by berndhahnebach** [comment]
that explains why it is not a problem ... the value is needed here because it is one item, but the value will not be used later ...

**Comment by berndhahnebach** [comment]
Ahh I missed reading the code. It would be used to create a FreeCAD property. It should be created too even if the referenced entry is missing. It should be mentioned that the entry is missing.

**Comment by berndhahnebach** [comment]
IMHO IfcOS should not raise a not handled exception. If this would be handled by IfcOpenShell we could put it in a try: except:  ...

https://github.com/IfcOpenShell/IfcOpenShell/issues/2811

**Comment by berndhahnebach** [comment]
this would work with uptodate IfcOpenShell ... 

replace this line by the following code ... https://github.com/yorikvanhavre/FreeCAD-NativeIFC/blob/e7bd20e6eedf3afd703ad122cc9996a7747b6a82/ifc_tools.py#L199

```
# info_ifcentity = ifcentity.get_info()
# the above raises an unhandled excption on corrupted ifc files
# thus workaround
attribs = []
for anumber in range(20):
    try:
        attr = ifcentity.attribute_name(anumber)
    except Exception:
        break
    # print(attr)
    attribs.append(attr)

info_ifcentity = {}
for attr in attribs:
    try:
        value = getattr(ifcentity, attr)
    except Exception as e:
        value = "Error: {}".format(e)
        # print(e)
    # print(value)
    info_ifcentity[attr] = value

for attr, value in info_ifcentity.items(): 
```

whats your opinion ...

**Comment by berndhahnebach** [comment]
this would fix the unhandled IfcOpenShell exception https://github.com/berndhahnebach/FreeCAD-NativeIFC/commit/529d055c44f24a5e3d68eb9c25be24b938fc8e94

We could temporary use this until this gets some fix ... https://github.com/IfcOpenShell/IfcOpenShell/issues/2811

**Comment by yorikvanhavre** [comment]
That seems like a good solution. But then I'd put that in a separate function (ex. get_attributes() ) so if this gets fixed we just switch back

**Comment by berndhahnebach** [comment]
same opinion here :-)

https://github.com/yorikvanhavre/FreeCAD-NativeIFC/pull/26 see comment I added another two small changes ...

**Comment by berndhahnebach** [comment]
may be you could jump in here ... https://github.com/IfcOpenShell/IfcOpenShell/issues/2811#issuecomment-1447839548

**Comment by yorikvanhavre** [comment]
Fixed with #26
